app/controllers/pieces_controller.rb

```diff
@@ -5,20 +5,30 @@ class PiecesController < ApplicationController
   def update
     x = params[:piece][:x_coordinate].to_i
     y = params[:piece][:y_coordinate].to_i
-    
-    if @piece.move_to! x, y
-      flash[:success] = "#{@piece.color.capitalize} #{@piece.type} move to X#{x}/Y#{y} was valid" if @piece.update_attributes(piece_params.merge(has_moved?: true))
+  
+    unless @game.is_finished?
+      if @piece.move_to! x, y
+        flash[:success] = "#{@piece.color.capitalize} #{@piece.type} move to X#{x}/Y#{y} was valid" if @piece.update_attributes(piece_params.merge(has_moved?: true))
+      else
+        flash[:danger] = "#{@piece.color.capitalize} #{@piece.type} move to X#{x}/Y#{y} was invalid"
+      end
+      
+      if @game.ends_by_checkmate? @piece.color
+        set_game_winner
+        
+        flash[:info] = "Checkmate! #{Player.find(@game.winning_player_id).email} has won this Game!"
+      else
+        color = @piece.color.eql?('white') ? "Black" : "White"
+        
+        flash[:warning] = "#{color} King is in check" if @game.in_check?(@piece.color)
+      end
+      
+      refresh_game
     else
-      flash[:danger] = "#{@piece.color.capitalize} #{@piece.type} move to X#{x}/Y#{y} was invalid"
+      flash[:info] = "No more moves permitted. The Game is Over."
+      
+      refresh_game
```

Comment:
There's a redundant 'refresh_game' method here (see line 26). I would use just one of these after the conditional since this method is ran regardless of the condition.